# Electromagnet Design

My goal is to make an electromagnet that can push a small but strong permanent magnet against gravity to a height of at least an inch when confined to one degree of freedom.

So far, I haven't had much success, but now I think I have all the basic understanding I'll need to do it. In this document, I lay out my design considerations.

## Background

I happened to find a video made by Prof. Eric Laithwaite, who was involved in inventing maglev train technology. In one of his demonstrations, he shows a DC electromagnet that seems to be able to do what I want: <https://youtu.be/OI_HFnNTfyU?t=199>

After seeing this I thought it probably was possible to make the electromagnet that I want.

Laithwaite's electromagnet is strong enough to push a permanent magnet several inches across a table. It's also strong enough to pick up a heavy iron bar on contact. I estimate that the iron bar weighs around 3-6 lbs. So I'm willing to bet that this electromagnet is more than capable of what I want to do.

The professor shows that if he takes out the iron core, the electromagnet becomes much weaker, demonstrating how important a ferromagnetic core really is. None of the electromagnets I've built so far have utilized a ferromagnetic core properly. At the time, I didn't understand magnetostatics well enough to even be sure that I needed a metal core. I did briefly try to use a screw as a core, but I recently realized that the metal in that screw is only barely ferromagnetic.

Laithwaite's electromagnet runs on about 1.5 A, which isn't a huge amount. My UPS can give that much current, so I may be able to get something within a similar order of magnitude of strength. Although I can't be sure how much voltage his coil requires, so I don't know its power requirements.

I don't plan to simply copy Laithwaite's exact setup. Of course I don't even know all the parameters of his setup. But further, I don't think his design is optimal or even meant to be optimal. I'd like to see how small I can make the electromagnet while generating the field strength that I want.

Since my previous attempts, I have read through the magnetostatics chapers of the Griffiths book a few times, thought through the logic, worked some of the problems, and let it bake in my head.

I believe I now understand all the required theory to build a pretty good model of an electromagnet so that I can tune its parameters and optimize a design.

## High level plan

The basic idea is to pass a current through a wire to generate a magnetic field and then use that field to magnetize a piece of ferromagnetic material. The magnetized material produces its own field which, depending on the chosen material, can be multiple times stronger than the wire's field.

For example, according to [this magnetization curve plot](https://en.wikipedia.org/wiki/File:Magnetization_curves.svg), if we apply an externally generated field of $H = 10 \text{ A/inch}$ to a piece of either cast iron or nickel, the metal will produce a field of about $B = 0.3 \text{ T}$. Converting the H-field to Teslas:


$$
H = 10 \text{ A/inch} = 400 \text{ A/m}
$$
$$
\rightarrow \quad
\mu_0 H = (4 \pi \times 10^{-7} \text{ N/A}^2) (400 \text{ A/m})
\approx 5 \times 10^{-4} \text{ T}
$$
$$
\mu_r = \frac{B}{\mu_0 H} \approx 600
$$

So in this particular case, the field of the magnetized metal is about 600 times stronger than the field generated by the electrical current. In other words, the relative permeability of iron and nickel in this region is 600. That's a huge difference. If that magnetization curve is correct, then the relationship between $H$ and $B$ for iron and nickel seems to be approximately linear in the region $B < 0.3 \text{ T}$, where the equation $B = 600 \mu_0 H$ should hold approximately true.

The most convenient geometry for the electromagnet, as far as I'm aware, is a cylindrical metal core with a solenoid wrapped around it. The solenoid of course may have more than one layer.

The parameters I'll have to consider are:

* Core dimensions and material
* Wire diameter
* Number of solenoid layers wrapped around the core
* Operating current and voltage

I'll have to make sure the operating power does not heat the wire too much.

## Solenoid design

The main part of the design is the solenoid. Whatever solenoid we come up with, we just fill the interior cavity with some kind of metal. The choice of metal material and dimensions will make a difference, but everything hinges on good solenoid design. We have to parameterize the solenoid and consider how changing the parameters affects the magnetic field strength.

The solenoid has cylindrical length $l$. The wire that the solenoid is made out of has a radius of $w$. Along the length of one layer of the solenoid, we can fit $n$ turns of wire.

$$
l = n \times (2 w)
\quad\rightarrow\quad
n = \frac{l}{2w}
$$

The empty interior cavity of the solenoid is a cylinder with radius $a$. If we have $m$ solenoid layers packed squarely (a pessimistic assumption, instead of assuming that I can actually achieve good circle packing) then the outer radius of the solenoid is $b$.

$$
b = a + m \times (2 w)
\quad\rightarrow\quad
m = \frac{b - a}{2 w}
$$

If we pass a current $I$ through this layered solenoid, what field is generated in the center of the solenoid? For that we can use Ampere's law.

$$
\oint \mathbf H \cdot d \mathbf l = I_\text{enc}
$$

We can use a rectangular Amperian loop of dimensions $l \times b$. One side of length $l$ goes along the radial axis of the solenoid. The other side of length $l$ is parallel with the first and goes along the outside of the solenoid. The two sides of length $b$ connect the other two sides. If we use the approximation that $l \gg b$, then the field is zero for all edges except the one that goes through the center of the solenoid, and the field is approximately uniform everywhere within the inner radius $a$.

$$
\oint \mathbf H \cdot d \mathbf l
= H l = n m I
$$

$$
\rightarrow\quad
H
= \frac{I}{l} n m
= \frac{I}{l} \frac{l}{2w} \frac{b - a}{2w}
$$

$$
H
= \frac{I (b - a)}{4w^2}
$$

Of course we won't actually have $l \gg b$, but this approximation should be good enough.

## Power requirements

In order to analyze the power requirements of the solenoid, we need to know how long the wire is. We can approximate this by imagining that we have $n \times m$ separate circular loops of wire. The circumference of one of these loops depends on which solenoid layer it is part of, since outer layers have a greater radius.

A loop in the $k$-th layer, where $k$ runs from 0 to $m-1$ (innermost to the outermost), has radius (from the center of the loop to the center of the wire):

$$
r_k
= a + w + k \times (2 w)
= a + w (1 + 2k)
$$

So a loop in the $k$-th layer then has a circumference (measured along the center of the wire):

$$
C_k = 2 \pi r_k
= 2 \pi \Bigl( a + w (1 + 2k) \Bigr)
$$

One entire layer just has $n$ of these loops, so the length of wire in layer $k$ is:

$$
L_k
= n C_k
= \frac{l}{2w} C_k
= \pi l \Bigl( \frac{a}{w} + 1 + 2k \Bigr)
$$

Then summing up the lengths of all $m$ layers, we have a total wire length of $L_w$, where:

$$
L_w
= \sum_{k=0}^{m-1} L_k
= \pi l \Biggl( m \Bigl(\frac{a}{w} + 1\Bigr) +  2 \sum_{k=0}^{m-1} k \Biggr)
$$

That summation $\Sigma_{k=0}^{m-1}k$ is just a [triangular number](https://en.wikipedia.org/wiki/Triangular_number), which has a well known algebraic representation $m (m-1)/2$.

$$
L_w
= \pi l \Biggl(
    m \Bigl(\frac{a}{w} + 1\Bigr)
    + m (m - 1)
\Biggr)
= \pi l m \Bigl(
    \frac{a}{w} + 1
    + m - 1
\Bigr)
$$

$$
= \pi l m \Bigl(
    \frac{a}{w} + m
\Bigr)
= \pi l \frac{b - a}{2w} \Bigl(
    \frac{a}{w} + \frac{b - a}{2w}
\Bigr)
= \pi l \frac{b - a}{2w} \Bigl(
    \frac{b + a}{2w}
\Bigr)
$$

$$
L_w
= \frac{\pi l (b - a)(b + a)}{4 w^2}
$$

Now that we know the length of the wire, we can calculate its resistance. I'm going to use copper wire, which has a resistivity of $\rho = 1.68 \times 10^{-8} \Omega \cdot \text{m}$. $A$ is the cross-sectional area of the wire.

$$
A = \pi w^2
$$

The resistance of a wire is given by:

$$
R
= \frac{\rho L_w}{A}
= \frac{\rho l (b - a)(b + a)}{4 w^4}
$$

If I want to get a current of $I$ through the wire, then I'll need to apply a voltage $V$.

$$
V
= I R
= \frac{\rho I l (b - a)(b + a)}{4 w^4}
$$

And the power required to run it is $P$.

$$
P = I^2 R
= \frac{\rho I^2 l (b - a)(b + a)}{4 w^4}
$$

## Heat management

In my earlier experiments, I made two different coils out of the same 30 AWG copper wire. I noticed that they had different maximum currents before heating up too much. I did not actually measure the temperature, I just felt it with my hand. I tried keeping the coils turned on for a long time to find a slightly warm but acceptably cool equilibrium temperature and then write down the current and voltage.

One of the coils, which had 2 $\Omega$ of resistance, started to get too hot if the power supply went above current 0.9 A, voltage 1.8 V, power 1.62 W. For the other coil, which was larger with resistance 4.6 $\Omega$, the heat limit was roughly at 0.8 A, 3.69 V, 3 W.
 
So increasing the wire length by a factor of about 2.3x increased the max power by about 1.8x.

I don't really know anything about thermodynamics, but I figure that the power limit at a certain temperature should scale approximately linearly with a coil's surface area, keeping all other factors the same, because more surface area allows more heat to escape.

$$
P_\text{limit} \propto A
$$


If we increase the total length of wire in the solenoid by some factor, we increase its volume by that same factor. If we increase the volume of the solenoid without changing its overall shape, then the volume and surface area are related by:

$$
A \propto V^{2/3}
\quad\rightarrow\quad
P_\text{limit} \propto V^{2/3}
$$

Using that relation, if two solenoids made of the same wire have roughly the same overall shape (i.e. both have similar ratios $a/l$ and $b/l$) and the total length of wire in one solenoid is 2.3x greater than the other, we should expect its power limit to be greater by a factor of approximately $(2.3)^{2/3} = 1.74$ x. That essentially lines up with the experimental result I mentioned above, where I saw the power limit increase by 1.8x. So even though I'm working with very little experimental data, I can be somewhat confident that the power limit is approximately proportional to the surface area.

To get a good approximation of the power limit of our parameterized solenoid model, we can use the surface area of a cylinder that is just big enough to enclose the entire solenoid, so it has height $l$ and radius $b$. But we'll cut out the area of the empty circular parts of the flat faces that have radius $a$, since that part of the solenoid is filled with air and will radiate significantly less heat than the surfaces that are made of copper wire. (Of course, when we put a metal core in the solenoid, it will probably solve that problem, but it's better to have a slightly pessimistic approximation.)

$$
A
= 2\pi b l +  2\pi b^2 - 2\pi a^2
= 2\pi (b l +  b^2 - a^2)
$$

We'll use a proportionality constant $\gamma$ to relate the surface area to the power limit. We can determine the value of the constant experimentally.

$$
P_\text{limit}
= \gamma A
= 2\pi \gamma (bl + b^2 - a^2)
$$

$$
\rightarrow\quad \gamma
= \frac{P_\text{limit}}{2 \pi (bl + b^2 - a^2)}
$$

To measure $\gamma$, I have two coils made of copper wire, but the wire gauges are different and the dimensions of the coils are different.

$$
l_1 = 1.9 \text{ cm}
\quad
l_2 = 1 \text{ cm}
$$
$$
a_1 = 0.4 \text{ cm}
\quad
a_2 = 0.4 \text{ cm}
$$
$$
b_1 = 1.25 \text{ cm}
\quad
b_2 = 0.95 \text{ cm}
$$

Coil 1 reaches its power limit at around:

$$
V_\text{limit(1)} = 5 \text{ V}
\quad
I_\text{limit(1)} = 0.137 \text{ A}
$$

So its resistance and power limit are:

$$
R_1
= \frac{V}{I}
= 36.5 \text{ }\Omega
$$

$$
P_\text{limit(1)}
= V I
= 0.685 \text{ W}
$$

Then to calculate $\gamma_1$:

In [1]:
l1 = 1.9 / 100
a1 = 0.4 / 100
b1 = 1.25 / 100
P_limit1 = 5 * 0.137
gamma1 = P_limit1 / (2 * 3.14159 * (b1 * l1 + b1 ** 2 - a1 ** 2))
gamma1

288.6068248954076

$$
\gamma_1 = 288 \text{ W/m}^2
$$

Coil 2 reaches its power limit at around:

$$
V_\text{limit(2)} = 0.7 \text{ V}
\quad
I_\text{limit(2)} = 0.466 \text{ A}
$$

So its resistance and power limit are:

$$
R_2
= \frac{V}{I}
= 1.5 \text{ }\Omega
$$

$$
P_\text{limit(2)}
= V I
= 0.326 \text{ W}
$$

In [2]:
l2 = 1 / 100
a2 = 0.4 / 100
b2 = 0.95 / 100
P_limit2 = 0.7 * 0.466
gamma2 = P_limit2 / (2 * 3.14159 * (b2 * l2 + b2 ** 2 - a2 ** 2))
gamma2

306.7437890013103

$$
\gamma_2 = 306 \text{ W/m}^2
$$

So just based on these two data points, I'll approximate $\gamma$ to be:

$$
\gamma \approx 300 \text{ W/m}^2
$$

Calculating $P_\text{limit}$ with this $\gamma$ should fairly reliably give me a soft limitation of power requirements, given that I want to be able to keep my solenoid turned on indefinitely without overheating.

## Core material

The core needs to be made of a material with high permeability and low coercivity.

Low coercivity means that the material does not retain much of an internal magnetic field when the external field is turned off. Coercivity is just the width of the hysteresis curve. In other words, it is the range of possible internal magnetic field strengths $B$ that the material could possibly produce when $H=0$. So the lower the coercivity, the more linear and reversible the relationship between $B$ and $H$ will be.

As far as I've found, an accessible and relatively cheap option with pretty good magnetic properties is 1018 cold rolled steel, which can be bought from [Metal Supermarkets](https://www.metalsupermarkets.com/product/mild-steel-round-bar-cold-rolled-1018/) in cylinders of a specified diameter and length.

According to [this plot](https://www.researchgate.net/figure/B-H-curve-for-1018-carbon-steel-A-Alers-and-Igarashi-26-W-Thompson-25-and-V_fig3_223314887), the relation between $B$ and $H$ is relatively linear around the origin. The relative permeability in the linear region is at least around:

$$
\mu_r
= \frac{B}{\mu_0 H}
= \frac{1.3 \text{ T}}{(4 \pi \times 10^{-7} \text{ N/A}^2) (5000 \text{ A/m})}
\approx 200
$$

There are other materials that would give better permeability, but I haven't found any affordable options for metal cylinders of arbitrary dimensions that beat 1018 steel.

Apparently 1006 or 1008 annealed steel would be a better option if I can find it. But it has to be annealed in order to get a better permeability. That can add a lot to the cost, and I'm having trouble finding a supplier.

1018 should be suitable enough for me, so I'm going with that for now unless I find something better later.

## Core design

The dimensions of the core affect the strength of the field at a distance from it. For instance, imagine we have two different solenoids of the same cylindrical length $l$ which both generate the same H-field strength in their respective cores and thus have the same $B$ within their cores. But now imagine that one of the cores has a cylindrical radius 10 times the radius of the other. The volume of the larger core is about $\pi 10^2 \approx 300$ times greater than the smaller one, so at a distance, the field strength will be around 300 times greater.

So the dimensions of the core significantly affect the strength of the field we can generate outside of the core.

We need to find an expression for the field that our cylindrical core generates at a distance from it, parameterized by its length $l$ and radius $a$. We already know how to determine the approximately uniform field $\mathbf H$ that the solenoid applies to the core. With relative permeability $\mu_r$, the core should produce an approximately uniform internal field:

$$
\mathbf B_\text{inside}
= B_\text{inside} \hat z
= \mu_0 \mu_r \mathbf H
$$

Since the field inside the core is uniform, it is uniformly magnetized.

$$
\mathbf M = M \hat z
$$

Let's find out how $M$ and $B_\text{inside}$ are related.

The field generated by a magnetized object is precisely the same as a field generated by a [magnetization current](https://en.wikipedia.org/wiki/Magnetization#Magnetization_current), which is made of a volume current $\mathbf J_b$ within the object and a surface current $\mathbf K_b$ on the surface of the object.

In our case, the volume current is:

$$
\mathbf J_b
= \nabla \times \mathbf M
= 0
$$

The surface current is:

$$
\mathbf K_b = \mathbf M \times \hat n
$$

The two flat surfaces of the cylindrical core have normals parallel to $\mathbf M$, so they contribute nothing to the surface current. But the curved surface is orthogonal to $\mathbf M$ everywhere, so it has nonzero surface current.

$$
\mathbf K_b
= \mathbf M \times \hat s
= M \hat \phi, \text{ for } s = a
$$

This is just a uniform surface current that flows counter-clockwise around the curved surface of the core.

So to find the field generated by the core, we can just treat it the same way we'd treat a solenoid. To relate $M$ and $B$, we can use Ampere's law on the surface current, much like we did earlier for the multi-layered solenoid, to approximate the field inside the core.

$$
\oint \mathbf B \cdot d \mathbf l
= \mu_0 I_\text{enc}
$$

$$
= B_\text{inside} l
= \mu_0 \int_0^l K_b dl
= \mu_0 l M
$$

$$
\rightarrow\quad
M
= \frac{B_\text{inside}}{\mu_0}
= \mu_r H
$$

We can approximate the field outside of the core with a magnetic dipole positioned at the center of the core. We just have to integrate the magnetization over the core's volume.

$$
\mathbf m
= \underset{V}{\int} \mathbf M d\tau
= \pi a^2 l \mathbf M
= \pi a^2 l \mu_r \mathbf H
$$

From Intro to Electrodynamics by Griffiths, if $\mathbf m$ is along $\hat z$, the field it generates is:

$$
\mathbf B_\text{outside} = \frac{\mu_0 m}{4 \pi r^3}(
  2 \cos\theta \hat r
  + \sin\theta \hat \theta
)
$$

Of course this is a far field approximation, which becomes more accurate the further we move away from core. If I want to maximize the equilibrium height of a permanent magnet suspened above the electromagnet, this approximation should suit me well.

Also, we should keep in mind that the dipole field is exactly equal to the field outside of a uniformly magnetized sphere. So the more closely our core resembles a sphere, the more accurate the dipole approximation is. In fact, if our cylindrical core has $2 a = l$, it should actually perform better than a spherical core of the same volume because we're effectively squishing more of the magnetized material upward so that it's closer to the suspended permanent magnet.

In any case, it's clear now that increasing the volume of the core, while keeping its shape and the H-field constant, will increase the strength of the field linearly.

## Equilibrium height

Now that we have a way to approximate the field $\mathbf B$ generated by the electromagnet, we can determine the height at which we can suspend a permanent magnet that has a known dipole moment $\mathbf m_p$ and mass $M$.

The electromagnet is sitting on a table with dipole moment oriented upward. The permanent magnet is at a height $z$ directly above the electromagnet's center ($z - l/2$ above the electromagnet's upper surface), with dipole moment oriented downward. Then the magnetic force on the permanent magnet is:

$$
\mathbf F_m
= \nabla (\mathbf m_p \cdot \mathbf B)
= \nabla \Bigl(
    (-m_p \hat z) \cdot (B_{\theta=0} \hat z) \Bigr)
= -\nabla \Bigl(
\frac{\mu_0 m m_p}{2 \pi z^3}
\Bigr)
$$

$$
\mathbf F_m
= \frac{3 \mu_0 m m_p}{2 \pi z^4} \hat z
$$

The magnetic force on the permanent magnet is:

$$
\mathbf F_g = - M g \hat z
$$

Let's find out what the height $z$ is when the two forces equally oppose each other.

$$
F_m = F_g
\quad\rightarrow\quad
M g = \frac{3 \mu_0 m m_p}{2 \pi z^4}
$$

$$
z
= \Bigl(\frac{3 \mu_0 m m_p}{2 \pi M g}\Bigr)^{1/4}
$$

It will be more important to know how high the magnet is suspended above the the electromagnet's top surface. That is just:

$$
h
= \Bigl(\frac{3 \mu_0 m m_p}{2 \pi M g}\Bigr)^{1/4} - \frac{l}{2}
$$

Of course if $h < 0$, then one of our magnets is not strong enough. I'd like to shoot for $h > 3 \text{ cm}$ if possible.

But now the question is, how can we figure out the moment $m_p$ of the permanent magnet? Turns out it's very easy to do it experimentally, as long as we have two of the same permanent magnet, and they are ring magnets. We can just stack them on a rod with opposing poles, measure the distance between them, and weigh them. In my case, I measure:

$$
z = 2.7 \text{cm} = 0.027 \text{ m}
\quad\quad
M = 5 \text{ g} = 0.005 \text{ kg}
$$

Then we just solve for $m_p$ and plug in all the known values.

$$
z^4
= \frac{3 \mu_0 {m_p}^2}{2 \pi M g}
$$

$$
m_p = z^2 \sqrt{\frac{2 \pi M g}{3 \mu_0}}
$$

In [3]:
import numpy as np
M = 0.005  # kg
g = 9.8   # m/s^2
mu_0 = 4 * np.pi * 1e-7  # N/A^2
z = 0.027  # m

m_p = z**2 * (2 * np.pi * M * g / (3 * mu_0))**0.5
print(f'{m_p:.3f} A m^2')

0.208 A m^2


So the magnetic moment of my permanent magnet rings is about:

$$
m_p \approx 0.208 \text{ A}\cdot\text{m}^2
$$

## Optimization

To summarize from the above, we've come up with a few equations to characterize the performance of the electromagnet.

$$
H
= \frac{I (b - a)}{4w^2}
\tag{1}
$$
$$
P
= \frac{\rho I^2 l (b - a)(b + a)}{4 w^4}
\tag{2}
$$
$$
P_\text{limit}
= 2\pi \gamma (bl + b^2 - a^2)
\tag{3}
$$
$$
m
= \pi a^2 l \mu_r H
\tag{4}
$$
$$
B_\text{outside} \propto m
\tag{5}
$$

I'm going to solve equation (2) for $I$ and plug it into equation (1), to get an equation for $B$ that depends only on the geometry of the coil ($l$, $a$, $b$, and $w$) and the power $P$ being supplied to it.

$$
I
= \sqrt{\frac{4 w^4 P}{\rho l (b - a)(b + a)}}
= 2 w^2 \sqrt{\frac{P}{\rho l (b - a)(b + a)}}
$$

$$
H
= \frac{b - a}{4w^2}
2 w^2 \sqrt{\frac{P}{\rho l (b - a)(b + a)}}
$$


$$
H = \frac{1}{2}
\sqrt{\frac{P (b - a)}{\rho l(b + a)}}
$$

(To be continued.)